# 7. 오류처리

> 1. 오류 코드보다 예외를 사용하라
> 2. Try-Cattch-Finally 문부터 작성하라
> 3. 미확인 예외를 사용하라
> 4. 예외에 의미를 제공하라
> 5. 호출자를 고려해 예외 클래스를 정의하라
> 6. 정상 흐름을 정의하라
> 7. null을 반환하지 마라
> 8. null을 전달하지 마라
> 9. 결론

## 1. 오류 코드보다 예외를 사용하라

오류가 발생한다면 예외를 던지는 편이 좋다. 그래야 호출자 코드가 깔끔해지기 때문이다. 그리고 논리가 오류 처리 코드와 뒤섞이지 않기 때문이다. 밑의 코드를 확인해보자 의외로 많이 더럽지 않음을 알 수가 있다

```java
public class DeviceController {
    ...
    
    public void sendShutDown() {
        try {
            tryToShutDown();
        } catch (DeviceShutDownError e) {
            logger.log(e);
        }
    }
    
    private void tryToShutDown() throws DeviceShutDownError {
        DeviceHandle handle = getHandle(DEV1);
        DeviceRecord record = retrieveDeviceRecord(handle);
        
        pauseDevice(handle);
        clearDeviceWorkQueue(hnadle);
        closeDevice(handle);
    }
    
    private DeviceHandle getHandle(DeviceId id) {
        ...
    }
}
```

## 2. Try-Catch-Finally 문부터 작성하라

어떻게 보면 try블록은 트랜잭션과 비슷하다. 아래 예제를 보자. 의식의 흐름과 함께 코드를 계속해서 수정해보자. 파일이 없으면 예외를 던지는지 알아보는 단위 테스트다

```java
@Test(expected = StorageException.class)
public void retrieveSectionShouldThrowOnInvalidFileName() {
    sectionStore.retrieveSection("invalid - file");
}
```

위 단위 테스트에 맞춰서 아래 코드를 구현해보았다.

```java
public List<RecordedGrip> retrieveSection(String sectionName) {
    // 실제로 구현할 때까지 비어있는 더미를 반환
    return new ArrayList<RecordedGrip>();
}
```

우리가 뭘 하고 있었나? 파일이 없으면 예외를 던지는지 알아보는 테스트이다. 즉, 예외를 던지지 않으면 테스트는 실패인 것이다. 그래서 위 코드를 조금 수정한다

```java
public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
        FileInputStream stream = new FileInputStream(sectionName);
    } catch (Exception e) {
        throw new StorageException("retrieval error", e);
    }
    return new ArrayList<RecordedGrip>();
}
```

예외를 던지니 이제 테스트가 성공할 것이다. 리펙토링도 조금 더 가능하다. catch에서 예외 유형을 좁히는 것이다. 아래와 같이 리펙토링도 조금 진행해보자

```java
public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
        FileInputStream stream = new FileInputStream(sectionName);
        stream.close();
    } catch (FileNotFoundException e) {
        throw new StorageException("retrieval error", e);
    }
    return new ArrayList<RecordedGrip>();
}
```

try-catch 구조로 범위를 정의했으니 TDD를 사용해 나머지 논리를 추가해줄 수가 있다.

## 3. 미확인 예외를 사용하라

확인된 예외는 멋질 수도 있다. 실제로 옛날에는 멋진 아이디어라고 생각했기 때문이다. 하지만 논란은 끝난지 오래다. 확인된 오류로 치르는 비용은 꽤 심각하다. 확인된 예외는 OCP를 위반한다. 메서드에서 확인된 예외를 던졌는데 catch블록이 세단계 위에 있다면 그 사이 메서드 모두가 선언부에 해당하는 예외를 정의해야한다. 즉, 하위 단계에서 코드를 변경하면 상위에 영향을 끼쳐 상위도 고쳐야한다.

즉, 우리는 예외를 확인하기 위해 굳이 코드를 계속 수정해야하고 확인된 예외에 따라 다른 결과를 보여준다. 문제가 조금 심각하다. 굳이 쓸 필요는 없지만 몇몇은 유용할 수 있다. 라이브러리에 의존하고 있다면 라이브러리에서 시키는 대로 해야하기 때문이다.

## 4. 예외에 의미를 제공하라

예외를 던질때 예외만 던지지 말 것! 전후상황을 알려줘야 문제원인 파악과 해결을 할 수가 있다.

## 5. 호출자를 고려해 예외 클래스를 정의하라

오류를 분류하는 방법은 아주 다양하다. 컴포넌트로 분류를 하던지, 위치로 분류하던지 매우 다양하게 분류가 가능하다. 그리고 우리는 이 분류를 형편없게 하면 안된다. 그리고 너무 천편일률적으로 예외에 대해 처리하지 말자. 꽤 많은 프로그래머가 다음 루트대로 처리할 것이다. `1) 오류를 기록한다. 2) 프로그램을 계속 수행해도 좋은지 판별한다`

호출하는 라이버르리 API를 감싸면서 예외 유형 하나를 반환하는 방식을 추구하자.

```java
LocalPort port = new LocalPort(12);
try {
    port.open();
} catch (ProtDeviceFaailure e) {
    reportError(e);
    logger.log(e.getMessage(), e);
} finally {
    ....
}
```

LoclPort 클래스는 단순히 ACMEPort 클래스가 던지는 예외를 잡아 변환하는 감싸기(wrapper) 클래스일 뿐이다

```java
public class LocalPort {
    private ACMEPort innerPort;
    
    public LocalPort(int portNumber) {
        innerPort = new ACMEPort(portNumber);
    }
    
    public void open() {
        try {
            innerPort.open();
        } catch (DeviceResponseException e) {
            throw new PortDeviceFailure(e);
        } catch (ATM1212UnlockedException e) {
            throw new PortDeviceFailure(e);
        } ....
    }
}
```

LocalPort클래스처럼 ACMEPort를 감싸는 클래스는 매우 유용하다. 외부 API는 감싸는 기법이 최선이다!

## 6. 정상 흐름을 정의하라

앞에서의 설명 모두 다 잘 충실히 이행한다면 비즈니스 논리와 오류 처리가 잘 분리된 코드를 이용할 수가 있을 것이다. 하지만 오류 코드가 구석으로 내팽켜치던가, 적절치 않게 에러를 반환한다거나 하는 일이 발생할 수도 있다. 

우리는 이 일에 특수 사례 패턴을 이용할 수가 있다. error를 dao로 고쳐 해당 객체를 반환할 수 있도록 하는 방식이다. [간단한 설명](https://harrislee.tistory.com/63)

## 7. null 을 반환하지 마라

오류 처리 후 자주 저지르는 것이 null을 반환하는 습관이다. 이 문제는 통제가 불가능해진다. 특수 사례 패턴을 이용하여 에러를 반환시키는 것도 괜찮은 방안이다. 에러를 반환 받은 곳에서 다른 로직을 취할 수도 있기 때문이다.

## 8. null을 전달하지 마라

메서드로 null을 전달하는 코드는 피하자. null은 우리가 제대로 된 상황파악이 어렵다. 에러가 일어난 것인가? 아니면 진짜 null을 필요해서 보낸 것인가. 우리가 통제하기 어려운 상황이기에 이 경우는 피하는 것이 좋을 것 같다.

## 9. 결론

깨끗한 코드이면서 잘돌아가기 까지 해야한다. 이것들은 상충된 것이 아니고 에러처리를 잘 한다면 잘 넘어갈 수가 있는 문제이다.